primitives/src/circuit/merkle_tree/sparse_merkle_tree.rs

```diff
@@ -0,0 +1,581 @@
+            .zip(merkle_proof.proof.iter().skip(1))
+            .filter_map(|(branch, node)| match node {
+                MerkleNode::Branch { value: _, children } => Some((children, branch)),
+                _ => None,
+            })
+            .map(|(children, branch)| {
+                Ok(Merkle3AryNodeVar {
+                    sibling1: self.create_variable(children[0].value())?,
+                    sibling2: self.create_variable(children[1].value())?,
+                    is_left_child: self.create_boolean_variable(branch == &0)?,
+                    is_right_child: self.create_boolean_variable(branch == &2)?,
+                })
+            })
+            .collect::<Result<Vec<Merkle3AryNodeVar>, CircuitError>>()?;
+
+        // `is_left_child`, `is_right_child` and `is_left_child+is_right_child` are
+        // boolean
+        for node in nodes.iter() {
+            // Boolean constrain `is_left_child + is_right_child` because a node
+            // can either be the left or the right child of its parent
+            let left_plus_right =
+                self.add(node.is_left_child.into(), node.is_right_child.into())?;
+            self.enforce_bool(left_plus_right)?;
+        }
+
+        let pos = self.create_variable(merkle_proof.pos.clone().into())?;
+
+        Ok(Self::NonMembershipProofVar {
+            node_vars: nodes,
+            pos_var: pos,
+        })
+    }
+}
+
+impl<F> MerkleTreeGadget<SparseMerkleTree<F>> for PlonkCircuit<F>
+where
+    F: RescueParameter,
+{
+    type MembershipProofVar = Merkle3AryMembershipProofVar;
+
+    fn create_membership_proof_variable(
+        &mut self,
+        merkle_proof: &MembershipProof<F>,
+    ) -> Result<Self::MembershipProofVar, CircuitError> {
+        MerkleTreeHelperGadget::<SparseMerkleTree<F>>::constrain_membership_proof(
+            self,
+            merkle_proof,
+        )
+    }
+
+    fn create_root_variable(&mut self, root: NodeVal<F>) -> Result<Variable, CircuitError> {
+        self.create_variable(root)
+    }
+
+    fn is_member(
+        &mut self,
+        leaf_idx_var: Variable,
+        proof_var: Self::MembershipProofVar,
+        root_var: Variable,
+    ) -> Result<BoolVar, CircuitError> {
+        let computed_root_var = MerkleTreeHelperGadget::<SparseMerkleTree<F>>::compute_merkle_root(
+            self,
+            leaf_idx_var,
+            &proof_var,
+        )?;
+        self.is_equal(root_var, computed_root_var)
+    }
+
+    fn enforce_membership_proof(
+        &mut self,
+        leaf_idx_var: Variable,
+        proof_var: Self::MembershipProofVar,
+        expected_root_var: Variable,
+    ) -> Result<(), CircuitError> {
+        let bool_val = MerkleTreeGadget::<SparseMerkleTree<F>>::is_member(
+            self,
+            leaf_idx_var,
+            proof_var,
+            expected_root_var,
+        )?;
+        self.enforce_true(bool_val.into())
+    }
+}
+
+trait SparseMerkleTreeHelperGadget<F: RescueParameter> {
```

Comment:
similar to my comment here: https://github.com/EspressoSystems/jellyfish/pull/172#discussion_r1069233944, I tend to prefer removing this trait and using a private stand-alone function instead of methods of `PlonkCircuit` instead.

Reply:
Done in https://github.com/EspressoSystems/jellyfish/pull/172/files/f2d2e8e0aa915c3025ac117c040a3b3ff93e2c36..1a266b75b17276b05eca60092d78ec07b02901b5